[WW-5141] Support for JakartaEE

jdyer1 opened the pull request:
WW-5141
[DRAFT - for evaluation only]

I am trying to help begin the effort to support jakarta ee 9+.  This PR represents a "least amount of work" path to the goal.  The code all compiles, but tests do not yet.  There are some BIG TODO's remaining also.   I would like to know if this is the correct path and what is thought to be the next steps completing this ticket.

This is a huge PR!  To help any reviewer to navigate, here is a summary of the changes.

1. modify references in java source:  "javax.">"jakarta."

2. pom dependency changes:
- Java version: 1.8 > 11
- javax.servlet-api > jakarta.servlet-api
- jsp-api > jakarta.servlet.jsp-api
- validation-api > jakarta.validation-api
- javax.el > jakarta.el
- cdi-api > jakarta.enterprise.cdi-api
- jstl > jakarta.servlet.jsp.jstl-api
- persistence-api > jakarta.persistence-api
- hibernate-validator.version: 6.2.4.Final > 8.0.1.Final
- spring.platformVersion: 6.2.4.Final > 8.0.1.Final
- tomcat-* version: 8.5.53 > 10.1.15
- sitemesh to g:org.sitemesh a:sitemesh v:3.2.0-M2  (TODO:  Maybe not neccessary???)
- spring-test version set to 4.3.0.RELEASE for plugins/cdi/pom.xml only.  See comment in pom.xml

3. Code changes

3a./plugins/embeddedjsp/../Generator.java
- change to: "out.printil("    extends jakarta..."

3b. plugins/embeddedjsp/../PageContextImpl.java
- JspException has no "getRootCause()".  change to "getCause()".\

3c. plugins/embeddedjsp/../JspCServletContext.java
- Add no-op methods for those added to interface

3d. /plugins/embeddedjsp/../JspServletWrapper.java
- Remove special logic for "instanceof SingleThreadModel" (removed from api)

3e. /plugins/json/../StrutsMockHttpServletRequest.java
- "com.mockobjects.servlet.MockHttpServletRequest" > "org.springframework.mock.web.MockHttpServletRequest"

3f. /plugins/json/../StrutsMockHttpSession.java
- "com.mockobjects.servlet.MockHttpServletRequest" > "org.springframework.mock.web.MockHttpServletRequest"

3g. /core/../RequestUtils.java
- request.getAttrubite("javax.servlet > ..jakarta.servlet

3h. /core/../Include.java
- request.getAttrubite("javax.servlet > ..jakarta.servlet

3i. /core/../ServletUrlRenderer.java
- request.getAttrubite("javax.servlet > ..jakarta.servlet

3j. /core/../FreemarkerTemplateEngine.java
- ClassLoaderUtil.loadClass("javax.servlet > ..jakarta.servlet

3k. /core/../DefaultDispatcherErrorHandler.java
- request.setAttribute("javax.servlet > ..jakarta.servlet

3l. /core/../Dispatcher.java
- servletContext.getAttribute("javax.servlet > ..jakarta.servlet

3m. /core/../StrutsRequestWrapper.java
- key.startsWith("jakarta.servlet > ..jakarta.servlet

3n. /core/../ServletDispatcherResult.java
- request.getAttribute("jakarta.servlet > ..jakarta.servlet

3o. /core/../FreemarkerManager.java
- request.getAttribute("jakarta.servlet > ..jakarta.servlet (2 places)

3p./core/../DefaultUrlHelper.java
- request.getAttribute("jakarta.servlet > ..jakarta.servlet

3q. /plugins/embeddedjsp/../Generator.java
- Several changes like: "out.print("jakarta.servlet"

3r. /plugins/embeddedjsp/../TagLibraryInfoImpl.java
- type = "javax.servlet. > ..jakarta.servlet

3s. /plugins/../JspRuntimeLibrary.java
- 4 places > "javax." > "jakarta."

3t. /plugins/embeddedjsp/../PageContextImpl.java
- many places > "javax." > "jakarta."

3u. /plugins/embeddedjsp/../JspServlet.java
- request.getAttribute("jakarta.servlet > ..jakarta.servlet (2 places)

3v. /plugins/embeddedjsp/../JspServletWrapper.java
- request.getAttribute("jakarta.servlet > ..jakarta.servlet

3w. /plugins/embeddedjsp/../LocalStrings*.properties
- many changes

3x. /plugins/../struts-osgi.properties
- javax > jakarta

3y. /plugins/tiles/../TilesDispatchServlet.jav
- request.getAttribute("jakarta.servlet > ..jakarta.servlet

3z. Removal of method from ElResolverImpl, removed from API.

4. Many Javadoc updates

5. Removals
- plugins/portlet-junit
- plugins/portlet-mocks
- plugins/portlet (not removed from struts2-core/src/site/resources/tags)
- plugins/portlet-tiles-plugin
- plugins/ddr
- plugins/pell-multipart

6. TODO

6a. freemarker
- revert temporary commit b8b9f8aa372adfdbbc72c3d71ffa204a93288fc0
- revert tempoaray commit 8d3880a6707a818f2d2a6dcde2b76a1612eb87c2

6b. velocity-tools
- revert temporary commit def5254a5744765735350fb04ff788c0784bc6ef
- revert temporary commit 1fc70af56ba1e7f2062fc957336c8321401eed39
- revert temporary commit 1fc70af56ba1e7f2062fc957336c8321401eed39
- revert temporary commit d5e766e43a14b6d4fcebd11fc7411c111e003fa6

6c. commons-upload
- revert temporary commit 3a4c1dd6f040c432081c0ed9cd8e5a20637efab1)
- new coordinates to not-backward-compatible version: g:org.apache.commons a:commons-fileupload2-jakarta v:2.0.0-M1

6d. Tests
- many tests need to be migrsted away from the old MockObjects library to compile

lukaszlenart commented:
This is astonishing work 🎉 Thanks for taking time and looking into it it. 

I was planning to prepare a dedicated branch to start working on the Struts 7.x, see this [thread](https://lists.apache.org/thread/lt6pfhzmv1jojd8f1q5mcnzp7d00wcdf) in the dev@ mailing list. And the very first step would be dropping all the deprecated plugins to simplify work.

lukaszlenart commented:
Also I plan to use JDK17 as base version for Struts 7.x

eschulma commented:
I would like to see JDK 17 rather than 11. Thank you for working on this!

Will the struts-bootsrap plugin still work? Hopefully yes.

jdyer1 commented:
@lukaszlenart   I have been checking that thread, and look forward to an official start to 7.x soon!   My worry here is this is a very big PR and I do not know how you will want to incorporate any of this into the official branch.

 I agree JDK 17 is a good idea.

@eschulma  I am not sure about the struts-bootstrap plugin.  Like everything else, someone needs to validate everything before a release is cut to be sure we aren't leaving things broken. 

To that end, the next task on this line of work I think is to get the tests to compile and run.  I was stuck on the old servlet api mocks from "com.mockobjects"  I was hoping that the Spring mocks would be a drop-in replacement, but many of the tests use methods that are only on "com.mockobjects".  I couldn't find source code for that library, so it was difficult to know what those methods do.

eschulma commented:
@jdyer1 when the time comes I am happy to help test the struts-bootstrap plugin -- our app will likely break it very quickly if there are any problems.

lukaszlenart commented:
@jdyer1 I will prepare the branch in a few days, and then we can consider how to merge/rebase changes from the master branch.

@eschulma we will have to release some RC versions first to allow other plugins (like jQuery or Bootstrap) to follow

jdyer1 commented:
I added several new commits to my fork.  Now all code compiles and all unit tests pass, with these 5 new caveats.

1. I ignored freemarker tests (REVERT these commits)
- 8d465c66a871cb5f7df4781c556f7d71e2932af1
- 1c43fd1dc77f0478823c668bca8a0a31f7e9dd6b
- 908d06483906c50d448795cf99547da9f91a9ef8

2. I ignored Commons-Upload tests (REVERT these commits)
- e4fa6e40ef99190a82e071900e937886e7f1a381
- f0df35f2e4d3fd1939f58a425729640a4d612ce9

3. I ignored "embedded jsp plugin" test.  Possibly remove this deprecated plugin?  Or, fix and REVERT this commit.
- 7f12f15d92177b5776255aee2e315e5cc55e9ffc

4. I Removed deprcated OVAL plugin.  Dependency "net.sf.oval" depends on javax.persistence, no "jakarta" compatible release available.

5. The "Struts 2 Showcase Webapp" integration test does not work.  I do not know how to fix it, but it seems the web server is not starting.  Probably "Struts 2 Rest Showcase Webapp" has the same issue.
-----
I think these are the next TODO's:

1. Fix common-uploads to work with newer version
2. Figure out what to do about freemarker/sitemesh (perhaps use eclipse transformer? https://github.com/eclipse/transformer)
3. Fix tests for "embedded jsp plugin" or remove this deprecated plugin
4. Fix the "Struts 2 Showcase Webapp" integration tests

If anyone can give input as to if you feel this is the right track, I would appreciate it.  I will continue to work on some of these things as my time allows.  It would be wonderful to also see some activity from seasoned Struts developers!  I think having a Jakarta EE 9+ version of Struts to be important as we all need to eventually migrate our applications to stay on supported runtimes.

lukaszlenart commented:
Astonish work 👏 and feel free to remove the `EmbeddedJSP` and `OVal` plugins.

tyme-dev commented:
@jdyer1 I've went ahead and got Commons-Upload tests working and converted over to Fileupload2-jakarta.
since your pr is the reference one, I've went ahead and did a pull request on your branch in order to get it over here. 

Only a few things were trip-ups :

- fileupload doesn't seem to be reporting actual size of files if they are larger than size limit, so i had to change error messages.
- setting threshold for writing file to disk, using -1 no longer seems to work, neither does 0. Got it working with a threshold of 1 though. 
- had to reorder exception instance check since FileSize exceptions have changed which exception classes they inherit

jdyer1 commented:
Thank you @tyme-dev  I merged your changes.

sepe81 commented on pom.xml:
```diff
@@ -792,50 +748,56 @@
-                <artifactId>jsp-api</artifactId>
-                <version>2.1</version>
+                <groupId>jakarta.servlet.jsp</groupId>
+				<artifactId>jakarta.servlet.jsp-api</artifactId> 
```
```suggestion
                <artifactId>jakarta.servlet.jsp-api</artifactId> 
```

tabs2spaces

sepe81 commented on pom.xml:
```diff
@@ -792,50 +748,56 @@
-                <version>2.1</version>
+                <groupId>jakarta.servlet.jsp</groupId>
+				<artifactId>jakarta.servlet.jsp-api</artifactId> 
+				<version>3.1.0</version>
```
```suggestion
                <version>3.1.0</version>
```

sepe81 commented on pom.xml:
```diff
@@ -946,38 +908,12 @@
             </dependency>
-
-            <dependency>
+			<dependency>
```
revert line

sepe81 commented on plugins/velocity/pom.xml:
```diff
@@ -54,7 +54,7 @@
             <scope>test</scope>
         </dependency>
-
+        
```
```suggestion

```

revert change

sepe81 commented on bundles/pom.xml:
```diff
@@ -51,9 +51,9 @@
-        <dependency>
-            <groupId>javax.servlet</groupId>
-            <artifactId>javax.servlet-api</artifactId>
+        <dependency>            
```
```suggestion
        <dependency>
```

trim trailing space

sepe81 commented on apps/showcase/pom.xml:
```diff
@@ -99,8 +94,8 @@
-            <groupId>javax.servlet</groupId>
-            <artifactId>javax.servlet-api</artifactId>
+            <groupId>jakarta.servlet</groupId>
+			<artifactId>jakarta.servlet-api</artifactId> 
```
```suggestion
            <artifactId>jakarta.servlet-api</artifactId> 
```

sepe81 reviewed (changes requested):
some minor formatting issues

jdyer1 commented:
I think I have solved the problems with Freemarker and Sitemesh by using the eclipse transformer maven plugin [1].  To make it work, I added two new projects that transform the jars to Jakarta EE 9+ compliant versions.  Downstream projects then depend on these new projects.  As far as I can tell, this works a lot like the shade plugin.

Unfortunately, eclipse is not able to recognize the transformed jars so it shows many compile errors.  I spent a lot of time today trying to find a good solution to no avail.  It seems IntelliJ has the same issue.  

As a workaround, I can select "Maven > Disable Workspace Resolution" on the struts2-core, struts2-tiles-plugin and struts2-sitemesg-plugin.  Clearly this is a bad permanent solution.

Perhaps someone reading this has encountered a similar situation before and they can improve on my work here and make it work seamlessly for everyone.  Seamless is the key especially for an open source project!

One final note, Struts uses Sitemesh2 which is stuck forever it seems on "javax.".  There is a Sitemesh3 that supports "Jakarta." but it seems the migration would be non-trivial.

-----
[1] https://github.com/eclipse/transformer/tree/main/maven-plugins/transformer-maven-plugin

jdyer1 commented:
Thank you @sepe81 for the review.  I think I have fixed all the formatting errors you pointed out.

tyme-dev commented:
> I think I have solved the problems with Freemarker and Sitemesh by using the eclipse transformer maven plugin [1]. To make it work, I added two new projects that transform the jars to Jakarta EE 9+ compliant versions. Downstream projects then depend on these new projects. As far as I can tell, this works a lot like the shade plugin.
> 
> Unfortunately, eclipse is not able to recognize the transformed jars so it shows many compile errors. I spent a lot of time today trying to find a good solution to no avail. It seems IntelliJ has the same issue.
> 
> As a workaround, I can select "Maven > Disable Workspace Resolution" on the struts2-core, struts2-tiles-plugin and struts2-sitemesg-plugin. Clearly this is a bad permanent solution.
> 
> Perhaps someone reading this has encountered a similar situation before and they can improve on my work here and make it work seamlessly for everyone. Seamless is the key especially for an open source project!
> 
> One final note, Struts uses Sitemesh2 which is stuck forever it seems on "javax.". There is a Sitemesh3 that supports "Jakarta." but it seems the migration would be non-trivial.
> 
> [1] https://github.com/eclipse/transformer/tree/main/maven-plugins/transformer-maven-plugin

can you give this link a look and see if it fixes the problem? 
https://blog.payara.fish/how-to-use-the-eclipse-transformer-project-to-convert-a-3rd-party-library 
looks like we should be able to use the transformer, and then put that artifact into local maven repo for build to pick up.

lukaszlenart commented:
@jdyer1 could you rebase your branch onto `release/struts-7-0-x`? Thanks in advance!

jdyer1 commented:
@lukaszlenart  I think I successfully rebased my branch on "release/struts-7-0-x".  I see however this PR still says it wants to merge to "apache:master".  I wonder if I need a new PR to make it want to merge to the release branch?  

Please double check me, I have not done a rebase like this before.

Also, I am not entirely sure how to get the struts showcase tests to pass.  It seems to want to spin up a server and make requests to it, but it might need some kind of conversion now that we are on JEE 9+?

jdyer1 commented:
@tyme-dev   Yes, I did see that blog before and it was helpful.  I followed instructions from this other, more detailed post [1].  (I did not do the additional step involving the shade plugin, because to my knowledge neither Freemarker nor Sitemesh2 have transitive dependencies that need to be converted.) 

To be clear here, the build does work correctly with `mvn clean install` at the command line.  The issue here is Eclipse IDE does not seem to want to use the local repo artifact but instead uses what is currently in the workspace.  Of course there is no code actually in the workspace.

-----
https://omnifish.ee/2023/08/04/upgrade-to-jakarta-ee-10-transform-incompatible-dependencies/

lukaszlenart commented:
@jdyer1 looks good, and switch base for this PR to `release/struts-7-0-x` and it show some conflicts, as far I understand is about deleted plugins so it should be easy to rebase again. About Showcase it uses Maven Jetty Plugin to run Jetty container, this plugin need to be updated to the latest version which supports JakaratEE and should be fine.

jdyer1 commented:
@lukaszlenart  Ok, I merged in the "release/struts-7-0-x" branch and now we have a "no conflicts" green check.

I also upgraded to the latest jetty-maven plugin.  Unfortunately it still fails.  It cannot seem to find the Freemarker classes, no doubt related to the Eclipse Transformer usage versus the actual Freemarker jar.  But I cannot figure out what is wrong.  I see in the Struts Showcase "target" directory a "struts2-showcase.war" file and there is a WEB-INF/lib and sure enough the "struts2-freemarker-jakarta.jar" is present.  Opening up this jar, everything appears to be packaged correctly.

But I am not so sure this is where the jetty-maven-plugin is getting its classpath? It doesn't seem to be running the web application contained in that .war file, does it?  (If it did, it would not be having classloading issues for Freemarker?)

lukaszlenart commented:
The plugin doesn't run the war by default, it uses target & src so you can edit JSP files and update them in-flight. Temporally we can use `mvn jetty:run-war` instead of `mvn jetty:run` when running the workflow.

jdyer1 commented:
@lukaszlenart  I believe I resolved most of the problems with "struts-showcase".  Before I saw your entire comment about adding a dependency exclusion, I instead added the 2 modified jars using the "extra classpath" Jetty feature.  This seems to work.  I also got rid of DWR and, for now, commented-out the velocity plugin.

@tyme-dev There is one test failure, however, `it.org.apache.struts2.showcase.FileUploadTest`.  Is this something you can look at?

jdyer1 commented:
@lukaszlenart   I think I have solved the problems with Velocity-Tools, once again, by using the Eclipse Transformer.  Besides this there is one remaining failing unit test (Ignored for now) in the Struts Showcase package around File Uploads.  I also would like to figure out a better way for IDEs to work seamlessly with these Eclipse Transformer jar files.

Do you think we are at a good place to merge this into the "release/struts-7-0-x" branch and close out this PR.  I think it will help with collaboration if it were in an official release branch.  I'd also like to know what the roadmap is for a first "alpha" release?

lukaszlenart commented:
@jdyer1 this looks very good, we can merge it and continue work on improving the other parts. Also I can prepare a new release, let say `M1` (this what Tomcat team is using)

kusalk commented:
My browser keeps freezing trying to load the diff - I'll have to load it in my IDE instead

Probably better to split this one up a bit anyway, maybe do the deprecated plugins removals in a separate preliminary PR?

tyme-dev commented on core/src/main/java/org/apache/struts2/dispatcher/multipart/JakartaMultiPartRequest.java:
```diff
@@ -186,13 +191,14 @@ protected ServletFileUpload createServletFileUpload(DiskFileItemFactory fac) {
+        fac.setBufferSize(1);
         if (saveDir != null) {
-            fac.setRepository(new File(saveDir));
+            fac.setPath(saveDir);
```
@jdyer1 so the test is failing here. 

i did make a mistake, this should be` fac.setPath(new File(saveDir))` i believe. But even when doing that, you're still not able to get the path back at all. 

This seems like a bug with the M1 release of fileupload2, I'm never able to get a path back for files made with the DiskFileItemFactory, but i could be overlooking something too.

jdyer1 replied:
Maybe this is a question for the commons-upload dev list?  I wonder, also, is it vital that Struts be able to tell the difference between a missing file and an empty file?  This is the only problem we are having here, correct?

lukaszlenart commented:
_time to rebase_

jdyer1 commented:
ok @lukaszlenart  this branch is now in sync with `release/struts-7-0-x`.  What are the next steps in getting this moved to the release branch?

kusalk commented:
Looks like the assembly job is failing due to some missed deletions - I threw up #799 to try and rectify

lukaszlenart commented:
@jdyer1 could you grant me access to your fork?

jdyer1 commented:
@kusalk I synced my fork again with "release/struts-7-0-x".  I also think I solved all the whitespace issues you were seeing.

@lukaszlenart I invited you as a "collaborator" on my fork.  I am not so skilled with github, but I think that will give you +w access to the fork once you accept.

lukaszlenart reviewed (commented):
Looks good to me 👏

jdyer1 commented:
@lukaszlenart  We still have a test failure `it.org.apache.struts2.showcase.FileUploadTest`, currently @Ignore'd.  It is odd to me that none of the file upload unit tests fail.   I wonder if we are checking something invalid here, that is, if it is trying to determine the difference between a missing file and a zero-length file?  If this is all that is lost, is this really functionality we need to support?

lukaszlenart commented:
I think you are right, after re-reviewing the related issues I'm confident we should drop support for _upload an empty file_ - it would be good to introduce a proper error message instead and that's all :) If you want you can do it in your patch or we can create a ticket and do it afterwards.

https://issues.apache.org/jira/browse/FILEUPLOAD-295
https://issues.apache.org/jira/browse/WW-5088

jdyer1 commented:
@lukaszlenart Thank you for the confirmation about  uploading a file.  I'd prefer we get a new ticket for dropping support for empty files.  I'm fairly new to the struts code base and I am not fully sure the best way to introduce a new error message, etc.

lukaszlenart commented:
Ok done [WW-5366](https://issues.apache.org/jira/browse/WW-5366)

lukaszlenart reviewed (approved):
LGTM 👏

kusalk commented on apps/showcase/pom.xml:
```diff
@@ -145,6 +146,48 @@
+                  <id>copy-dependencies</id>
+                  <phase>package</phase>
+                  <goals>
+                    <goal>copy</goal>
```
I'm probably missing something here but why do we need to explicitly copy these jars?

jdyer1 replied:
@kusalk  I struggled when doing this to figure out how the jetty-maven-plugin was building its classpath.  It was not including the 4 jars we generate from the Eclipse Transformer, so the showcase integration tests all would fail.  The solution I found was to use the Jetty "extraclasspath" feature (see same pom.xml lines 236ff) in conjunction with copying the jars as you see here.  In my ignorance I may have passed up a more-elegant solution!

lukaszlenart replied:
It should be possible to play with `<exclusion>` and explicitly defining the re-packaged jakarta libs, but this can be done after merge

yasserzamani commented on core/src/test/java/org/apache/struts2/views/jsp/BeanTagTest.java:
```diff
@@ -51,8 +51,8 @@ public void testSimple() {
             fail();
         }
 
-        request.verify();
-        pageContext.verify();
```
Doesn't removing them skip some verifications that might fail with these changes? sorry I don't remember.

jdyer1 replied:
@yasserzamani Thank you for your review.  I can say for sure that "pageContext.verify()" is a no-op.  As for the request, my belief is the mock is designed to let the tester set up a series of expectations, then call "verify" when the test is ready to check them.  However, this test seems to set no expectations, so my belief is the "verify" method here also would do nothing.

You touch here on what I consider to be the most risky of these changes:  The "mockobjects" library for the servlet api could no longer be used so I replaced them with the Spring library's mocks.  However, the Spring mocks do not support this "set/verify" pattern so in the few places our code uses it, I had to make a judgement as to what I thought the original intent was.  I cannot say for sure I got this right 100% of the time!

yasserzamani replied:
Thanks for clarification @jdyer1 ! Yes you're right! I also couldn't find their usage. Those lines are old from 2006. Maybe underlying dependencies get updated making these left behind and why they don't make sense anymore.

yasserzamani reviewed (approved):
wow! a lot of works, thanks! I saw no changes in tests functionalities except in FileUploadTest which is already clarified and a few removements of `.verify()`s which I wasn't sure about. So LGTM as all tests still are passing :) 💯

sepe81 commented on core/pom.xml:
```diff
@@ -89,9 +89,8 @@
                 </configuration>
             </plugin>
-        </plugins>
+          </plugins>
```
```suggestion
        </plugins>
```

revert indentation

sepe81 commented on core/src/main/java/org/apache/struts2/dispatcher/multipart/JakartaMultiPartRequest.java:
```diff
@@ -72,16 +73,16 @@ public void parse(HttpServletRequest request, String saveDir) throws IOException
+            } else if (e instanceof FileUploadSizeException) {
+                FileUploadSizeException ex = (FileUploadSizeException) e;
+                errorMessage = buildErrorMessage(e, new Object[]{ex.getPermitted(), ex.getActualSize()});
+            }   else {
```
```suggestion
            } else {
```

revert indentation

sepe81 commented on core/src/main/java/org/apache/struts2/dispatcher/multipart/JakartaMultiPartRequest.java:
```diff
@@ -356,7 +362,11 @@ public void cleanUp() {
+                    try {
+                        item.delete();
+                    } catch (IOException e) {
+                        throw new RuntimeException(e);
```
Maybe use more specific https://docs.oracle.com/javase/8/docs/api/java/io/UncheckedIOException.html ?

sepe81 commented on core/src/test/java/org/apache/struts2/dispatcher/mapper/Restful2ActionMapperTest.java:
```diff
@@ -61,11 +62,11 @@ public Configuration getConfiguration() {
-        req.setupGetMethod("GET");
+        req.setRequestURI("/my/namespace/foo/");
+        req.setServletPath("/my/namespace/foo/");
+        //
```
remove empty comments or is something missing?

sepe81 commented on core/src/test/java/org/apache/struts2/interceptor/FileUploadInterceptorTest.java:
```diff
@@ -550,6 +553,7 @@ private String encodeTextFile(String bondary, String endline, String name, Strin
         sb.append(name);
         sb.append("\"; filename=\"");
         sb.append(filename);
+        sb.append("\"");
```
👍

sepe81 commented on core/src/test/java/org/apache/struts2/views/jsp/BeanTagTest.java:
```diff
@@ -83,8 +83,8 @@ public void testSimple_clearTagStateSet() {
 
-        request.verify();
-        pageContext.verify();
+        
```
remove two empty lines

sepe81 reviewed (approved):
LGMT, great work, maybe just some small fixups

jdyer1 commented:
Thank you @sepe81 for your review.  I pushed fixes for each of the problems you identified.

lukaszlenart commented:
Let's roll 🎉  Thanks a lot @jdyer1 for you work and all reviewers 👏

ClancyDamon commented:
I'm having some trouble building the 7-0-x release. Compilation seems to be okay, but it fails when executing the tests inside of the core module. Specifically org.apache.struts2.interceptor.ActionFileUploadInterceptorTest and org.apache.struts2.interceptor.FileUploadInterceptorTest.

`-------------------------------------------------------------------------------
Test set: org.apache.struts2.interceptor.ActionFileUploadInterceptorTest
-------------------------------------------------------------------------------
Tests run: 15, Failures: 2, Errors: 0, Skipped: 0, Time elapsed: 0.267 s <<< FAILURE! - in org.apache.struts2.interceptor.ActionFileUploadInterceptorTest
org.apache.struts2.interceptor.ActionFileUploadInterceptorTest.testMultipleAccept  Time elapsed: 0.021 s  <<< FAILURE!
junit.framework.AssertionFailedError
	at junit.framework.Assert.fail(Assert.java:55)
	at junit.framework.Assert.assertTrue(Assert.java:22)
	at junit.framework.Assert.assertNotNull(Assert.java:256)
	at junit.framework.Assert.assertNotNull(Assert.java:248)
	at junit.framework.TestCase.assertNotNull(TestCase.java:391)
	at org.apache.struts2.interceptor.ActionFileUploadInterceptorTest.testMultipleAccept(ActionFileUploadInterceptorTest.java:360)
	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:77)
	at java.base/jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.base/java.lang.reflect.Method.invoke(Method.java:568)
	at junit.framework.TestCase.runTest(TestCase.java:177)
	at junit.framework.TestCase.runBare(TestCase.java:142)
	at junit.framework.TestResult$1.protect(TestResult.java:122)
	at junit.framework.TestResult.runProtected(TestResult.java:142)
	at junit.framework.TestResult.run(TestResult.java:125)
	at junit.framework.TestCase.run(TestCase.java:130)
	at junit.framework.TestSuite.runTest(TestSuite.java:241)
	at junit.framework.TestSuite.run(TestSuite.java:236)
	at org.junit.internal.runners.JUnit38ClassRunner.run(JUnit38ClassRunner.java:90)
	at org.junit.runners.Suite.runChild(Suite.java:128)
	at org.junit.runners.Suite.runChild(Suite.java:27)
	at org.junit.runners.ParentRunner$4.run(ParentRunner.java:331)
	at org.junit.runners.ParentRunner$1.schedule(ParentRunner.java:79)
	at org.junit.runners.ParentRunner.runChildren(ParentRunner.java:329)
	at org.junit.runners.ParentRunner.access$100(ParentRunner.java:66)
	at org.junit.runners.ParentRunner$2.evaluate(ParentRunner.java:293)
	at org.junit.runners.ParentRunner$3.evaluate(ParentRunner.java:306)
	at org.junit.runners.ParentRunner.run(ParentRunner.java:413)
	at org.apache.maven.surefire.junitcore.JUnitCore.run(JUnitCore.java:49)
	at org.apache.maven.surefire.junitcore.JUnitCoreWrapper.createRequestAndRun(JUnitCoreWrapper.java:120)
	at org.apache.maven.surefire.junitcore.JUnitCoreWrapper.executeEager(JUnitCoreWrapper.java:95)
	at org.apache.maven.surefire.junitcore.JUnitCoreWrapper.execute(JUnitCoreWrapper.java:75)
	at org.apache.maven.surefire.junitcore.JUnitCoreWrapper.execute(JUnitCoreWrapper.java:69)
	at org.apache.maven.surefire.junitcore.JUnitCoreProvider.invoke(JUnitCoreProvider.java:146)
	at org.apache.maven.surefire.booter.ForkedBooter.runSuitesInProcess(ForkedBooter.java:385)
	at org.apache.maven.surefire.booter.ForkedBooter.execute(ForkedBooter.java:162)
	at org.apache.maven.surefire.booter.ForkedBooter.run(ForkedBooter.java:507)
	at org.apache.maven.surefire.booter.ForkedBooter.main(ForkedBooter.java:495)

org.apache.struts2.interceptor.ActionFileUploadInterceptorTest.testUnacceptedNumberOfFiles  Time elapsed: 0.019 s  <<< FAILURE!
junit.framework.ComparisonFailure: expected:<[Request exceeded allowed number of files! Permitted number of files is: 3!]> but was:<[Error uploading: {0}]>
	at junit.framework.Assert.assertEquals(Assert.java:100)
	at junit.framework.Assert.assertEquals(Assert.java:107)
	at junit.framework.TestCase.assertEquals(TestCase.java:260)
	at org.apache.struts2.interceptor.ActionFileUploadInterceptorTest.testUnacceptedNumberOfFiles(ActionFileUploadInterceptorTest.java:397)
	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:77)
	at java.base/jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.base/java.lang.reflect.Method.invoke(Method.java:568)
	at junit.framework.TestCase.runTest(TestCase.java:177)
	at junit.framework.TestCase.runBare(TestCase.java:142)
	at junit.framework.TestResult$1.protect(TestResult.java:122)
	at junit.framework.TestResult.runProtected(TestResult.java:142)
	at junit.framework.TestResult.run(TestResult.java:125)
	at junit.framework.TestCase.run(TestCase.java:130)
	at junit.framework.TestSuite.runTest(TestSuite.java:241)
	at junit.framework.TestSuite.run(TestSuite.java:236)
	at org.junit.internal.runners.JUnit38ClassRunner.run(JUnit38ClassRunner.java:90)
	at org.junit.runners.Suite.runChild(Suite.java:128)
	at org.junit.runners.Suite.runChild(Suite.java:27)
	at org.junit.runners.ParentRunner$4.run(ParentRunner.java:331)
	at org.junit.runners.ParentRunner$1.schedule(ParentRunner.java:79)
	at org.junit.runners.ParentRunner.runChildren(ParentRunner.java:329)
	at org.junit.runners.ParentRunner.access$100(ParentRunner.java:66)
	at org.junit.runners.ParentRunner$2.evaluate(ParentRunner.java:293)
	at org.junit.runners.ParentRunner$3.evaluate(ParentRunner.java:306)
	at org.junit.runners.ParentRunner.run(ParentRunner.java:413)
	at org.apache.maven.surefire.junitcore.JUnitCore.run(JUnitCore.java:49)
	at org.apache.maven.surefire.junitcore.JUnitCoreWrapper.createRequestAndRun(JUnitCoreWrapper.java:120)
	at org.apache.maven.surefire.junitcore.JUnitCoreWrapper.executeEager(JUnitCoreWrapper.java:95)
	at org.apache.maven.surefire.junitcore.JUnitCoreWrapper.execute(JUnitCoreWrapper.java:75)
	at org.apache.maven.surefire.junitcore.JUnitCoreWrapper.execute(JUnitCoreWrapper.java:69)
	at org.apache.maven.surefire.junitcore.JUnitCoreProvider.invoke(JUnitCoreProvider.java:146)
	at org.apache.maven.surefire.booter.ForkedBooter.runSuitesInProcess(ForkedBooter.java:385)
	at org.apache.maven.surefire.booter.ForkedBooter.execute(ForkedBooter.java:162)
	at org.apache.maven.surefire.booter.ForkedBooter.run(ForkedBooter.java:507)
	at org.apache.maven.surefire.booter.ForkedBooter.main(ForkedBooter.java:495)

`

`-------------------------------------------------------------------------------
Test set: org.apache.struts2.interceptor.FileUploadInterceptorTest
-------------------------------------------------------------------------------
Tests run: 15, Failures: 2, Errors: 0, Skipped: 0, Time elapsed: 0.207 s <<< FAILURE! - in org.apache.struts2.interceptor.FileUploadInterceptorTest
org.apache.struts2.interceptor.FileUploadInterceptorTest.testMultipleAccept  Time elapsed: 0.016 s  <<< FAILURE!
junit.framework.AssertionFailedError: expected:<3> but was:<0>
	at junit.framework.Assert.fail(Assert.java:57)
	at junit.framework.Assert.failNotEquals(Assert.java:329)
	at junit.framework.Assert.assertEquals(Assert.java:78)
	at junit.framework.Assert.assertEquals(Assert.java:234)
	at junit.framework.Assert.assertEquals(Assert.java:241)
	at junit.framework.TestCase.assertEquals(TestCase.java:384)
	at org.apache.struts2.interceptor.FileUploadInterceptorTest.testMultipleAccept(FileUploadInterceptorTest.java:373)
	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:77)
	at java.base/jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.base/java.lang.reflect.Method.invoke(Method.java:568)
	at junit.framework.TestCase.runTest(TestCase.java:177)
	at junit.framework.TestCase.runBare(TestCase.java:142)
	at junit.framework.TestResult$1.protect(TestResult.java:122)
	at junit.framework.TestResult.runProtected(TestResult.java:142)
	at junit.framework.TestResult.run(TestResult.java:125)
	at junit.framework.TestCase.run(TestCase.java:130)
	at junit.framework.TestSuite.runTest(TestSuite.java:241)
	at junit.framework.TestSuite.run(TestSuite.java:236)
	at org.junit.internal.runners.JUnit38ClassRunner.run(JUnit38ClassRunner.java:90)
	at org.junit.runners.Suite.runChild(Suite.java:128)
	at org.junit.runners.Suite.runChild(Suite.java:27)
	at org.junit.runners.ParentRunner$4.run(ParentRunner.java:331)
	at org.junit.runners.ParentRunner$1.schedule(ParentRunner.java:79)
	at org.junit.runners.ParentRunner.runChildren(ParentRunner.java:329)
	at org.junit.runners.ParentRunner.access$100(ParentRunner.java:66)
	at org.junit.runners.ParentRunner$2.evaluate(ParentRunner.java:293)
	at org.junit.runners.ParentRunner$3.evaluate(ParentRunner.java:306)
	at org.junit.runners.ParentRunner.run(ParentRunner.java:413)
	at org.apache.maven.surefire.junitcore.JUnitCore.run(JUnitCore.java:49)
	at org.apache.maven.surefire.junitcore.JUnitCoreWrapper.createRequestAndRun(JUnitCoreWrapper.java:120)
	at org.apache.maven.surefire.junitcore.JUnitCoreWrapper.executeEager(JUnitCoreWrapper.java:95)
	at org.apache.maven.surefire.junitcore.JUnitCoreWrapper.execute(JUnitCoreWrapper.java:75)
	at org.apache.maven.surefire.junitcore.JUnitCoreWrapper.execute(JUnitCoreWrapper.java:69)
	at org.apache.maven.surefire.junitcore.JUnitCoreProvider.invoke(JUnitCoreProvider.java:146)
	at org.apache.maven.surefire.booter.ForkedBooter.runSuitesInProcess(ForkedBooter.java:385)
	at org.apache.maven.surefire.booter.ForkedBooter.execute(ForkedBooter.java:162)
	at org.apache.maven.surefire.booter.ForkedBooter.run(ForkedBooter.java:507)
	at org.apache.maven.surefire.booter.ForkedBooter.main(ForkedBooter.java:495)

org.apache.struts2.interceptor.FileUploadInterceptorTest.testUnacceptedNumberOfFiles  Time elapsed: 0.017 s  <<< FAILURE!
junit.framework.ComparisonFailure: expected:<[Request exceeded allowed number of files! Permitted number of files is: 3!]> but was:<[Error uploading: {0}]>
	at junit.framework.Assert.assertEquals(Assert.java:100)
	at junit.framework.Assert.assertEquals(Assert.java:107)
	at junit.framework.TestCase.assertEquals(TestCase.java:260)
	at org.apache.struts2.interceptor.FileUploadInterceptorTest.testUnacceptedNumberOfFiles(FileUploadInterceptorTest.java:427)
	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:77)
	at java.base/jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.base/java.lang.reflect.Method.invoke(Method.java:568)
	at junit.framework.TestCase.runTest(TestCase.java:177)
	at junit.framework.TestCase.runBare(TestCase.java:142)
	at junit.framework.TestResult$1.protect(TestResult.java:122)
	at junit.framework.TestResult.runProtected(TestResult.java:142)
	at junit.framework.TestResult.run(TestResult.java:125)
	at junit.framework.TestCase.run(TestCase.java:130)
	at junit.framework.TestSuite.runTest(TestSuite.java:241)
	at junit.framework.TestSuite.run(TestSuite.java:236)
	at org.junit.internal.runners.JUnit38ClassRunner.run(JUnit38ClassRunner.java:90)
	at org.junit.runners.Suite.runChild(Suite.java:128)
	at org.junit.runners.Suite.runChild(Suite.java:27)
	at org.junit.runners.ParentRunner$4.run(ParentRunner.java:331)
	at org.junit.runners.ParentRunner$1.schedule(ParentRunner.java:79)
	at org.junit.runners.ParentRunner.runChildren(ParentRunner.java:329)
	at org.junit.runners.ParentRunner.access$100(ParentRunner.java:66)
	at org.junit.runners.ParentRunner$2.evaluate(ParentRunner.java:293)
	at org.junit.runners.ParentRunner$3.evaluate(ParentRunner.java:306)
	at org.junit.runners.ParentRunner.run(ParentRunner.java:413)
	at org.apache.maven.surefire.junitcore.JUnitCore.run(JUnitCore.java:49)
	at org.apache.maven.surefire.junitcore.JUnitCoreWrapper.createRequestAndRun(JUnitCoreWrapper.java:120)
	at org.apache.maven.surefire.junitcore.JUnitCoreWrapper.executeEager(JUnitCoreWrapper.java:95)
	at org.apache.maven.surefire.junitcore.JUnitCoreWrapper.execute(JUnitCoreWrapper.java:75)
	at org.apache.maven.surefire.junitcore.JUnitCoreWrapper.execute(JUnitCoreWrapper.java:69)
	at org.apache.maven.surefire.junitcore.JUnitCoreProvider.invoke(JUnitCoreProvider.java:146)
	at org.apache.maven.surefire.booter.ForkedBooter.runSuitesInProcess(ForkedBooter.java:385)
	at org.apache.maven.surefire.booter.ForkedBooter.execute(ForkedBooter.java:162)
	at org.apache.maven.surefire.booter.ForkedBooter.run(ForkedBooter.java:507)
	at org.apache.maven.surefire.booter.ForkedBooter.main(ForkedBooter.java:495)

`

I'm utilizing Amazon Corretto Java 17 (jdk 17.0.3_6). Is there something special I need to do, or some additional repositories I should get? I don't think this is related to earlier conversation about empty files, the tests that failed relate to the number of files uploaded.

lukaszlenart commented:
How do you run these tests? And maybe it's related to Amazon Corretto yet I don't have version locally to test

lukaszlenart commented:
When using Maven you should be able download SNAPSHOTs from ASF repo or you can download nigthlies from here https://struts.apache.org/builds.html#nightly-builds

ClancyDamon commented:
Sorry for the delay. Holidays and personal stuff.

I'm not doing anything special. I'm just running the maven install command from the source directory after downloading to my local machine - mvn -DskipAssembly=true install

I'm following the build instructions from apache - https://struts.apache.org/builds.html

lukaszlenart commented:
I would try first with `mvn -DskipAssembly package` or install Jakarta modules first
```
cd jakarta
mvn -DskipAssembly install
```
and then install the whole project. I noticed some problem with having these transformed modules next to the core modules.
